sdk/lib/locks/locks.cc

```diff
@@ -63,6 +52,17 @@ namespace
 				{
 					return 0;
 				}
+
+				// Do not attempt to lock if the destruct mode
+				// flag is set. Note that the lock is in
+				// destruction mode iff it is already locked,
```

Comment:
I don't think that's the reason: we are not setting the locked-for-destruction bit in our expected value, so the CAS will always fail if the lock-for-destruction bit is set.  Being locked for destruction is unusual (it should happen exactly once for any lock) and so we don't bother to check for it unless we've failed to acquire the lock.